From 166cd2d95101aecf50f7b0984f1b57bc23a31921 Mon Sep 17 00:00:00 2001
From: Sylvain Ageneau <ageneau@gmail.com>
Date: Wed, 19 Dec 2012 14:45:06 -0300
Subject: [PATCH 1/2] Fix GC_init on Android 4.0.3 and 4.1

---
 os_dep.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/os_dep.c b/os_dep.c
index 2730205..cacac24 100644
--- a/os_dep.c
+++ b/os_dep.c
@@ -436,7 +436,7 @@ GC_INNER char * GC_get_maps(void)
   /* for recent Linux versions.  This seems to be the easiest way to    */
   /* cover all versions.                                                */
 
-# if defined(LINUX) || defined(HURD)
+# if (defined(LINUX) || defined(HURD)) && ! defined(PLATFORM_ANDROID)
     /* Some Linux distributions arrange to define __data_start.  Some   */
     /* define data_start as a weak symbol.  The latter is technically   */
     /* broken, since the user program may define data_start, in which   */
@@ -455,7 +455,7 @@ GC_INNER char * GC_get_maps(void)
 
   GC_INNER void GC_init_linux_data_start(void)
   {
-#   if (defined(LINUX) || defined(HURD)) && !defined(IGNORE_PROG_DATA_START)
+#   if (defined(LINUX) || defined(HURD)) && !(defined(IGNORE_PROG_DATA_START) || defined(PLATFORM_ANDROID))
       /* Try the easy approaches first: */
       if ((ptr_t)__data_start != 0) {
           GC_data_start = (ptr_t)(__data_start);
-- 
1.8.4

