From e748f1ed58741152793e07a47cf71c8b2901f403 Mon Sep 17 00:00:00 2001
From: Sylvain Ageneau <ageneau@gmail.com>
Date: Sat, 4 Aug 2012 17:06:50 -0300
Subject: [PATCH 08/11] gmp

---
 configure_gmp_cross | 23 ++++++++++++++++++++---
 1 file changed, 20 insertions(+), 3 deletions(-)

diff --git a/configure_gmp_cross b/configure_gmp_cross
index cd8022f..9d11307 100755
--- a/configure_gmp_cross
+++ b/configure_gmp_cross
@@ -16,6 +16,9 @@ platform=$1
 . scripts.conf
 GMP_INSTALL_DIR=$GMP_INSTALL_ROOT_DIR/$platform
 srcdir=$ECLROOT/src/gmp
+#srcdir=$ECLROOT/../mpir-2.5.1
+host=$(${srcdir}/config.guess)
+
 builddir=$ECLROOT/build/$platform/gmp
 
 . $ECLROOT/platforms/$platform/config.sh
@@ -25,20 +28,34 @@ if [ ! -d ${builddir} ] ; then
     mkdir -p ${builddir}
 fi
 
+if grep -q MPIR $srcdir/Changelog; then
+    echo "Using MPIR library"
+    base_config_opt="--enable-gmpcompat"
+fi
 
 base_config_opts="\
         --disable-shared \
-	--with-pic"
+	${base_config_opt}"
 
 case $platform in
     host)
+	base_config_opts="\
+	--with-pic \
+	${base_config_opt}"
+
+	# MPIR-2.5.1: must use applenopic on sandybridge
+	if [[ "$host" == "sandybridge-apple-darwin"* ]]; then 
+	    host=core2-apple-darwin
+	fi
+
 	export CFLAGS="-m32"
 	export LDFLAGS="-m32"
 	export ABI=32
 	;;
-    iPhoneSimulator)
+    host64)
 	base_config_opts="\
-        --disable-shared"
+	--with-pic \
+	${base_config_opt}"
 	;;
 esac
 
-- 
1.8.0

